library/core/src/iter/sources/repeat_n.rs

```diff
@@ -74,31 +73,60 @@ pub fn repeat_n<T: Clone>(element: T, count: usize) -> RepeatN<T> {
 ///
 /// This `struct` is created by the [`repeat_n()`] function.
 /// See its documentation for more.
-#[derive(Clone, Debug)]
 #[stable(feature = "iter_repeat_n", since = "1.82.0")]
 pub struct RepeatN<A> {
     count: usize,
-    // Invariant: has been dropped iff count == 0.
-    element: ManuallyDrop<A>,
+    // Invariant: uninit iff count == 0.
+    element: MaybeUninit<A>,
 }
 
 impl<A> RepeatN<A> {
+    /// Returns the element if it hasn't been dropped already.
+    fn element_ref(&self) -> Option<&A> {
+        if self.count > 0 {
+            // SAFETY: The count is non-zero, so it must be initialized.
+            Some(unsafe { self.element.assume_init_ref() })
+        } else {
+            None
+        }
+    }
     /// If we haven't already dropped the element, return it in an option.
     ///
     /// Clears the count so it won't be dropped again later.
     #[inline]
     fn take_element(&mut self) -> Option<A> {
         if self.count > 0 {
             self.count = 0;
+            let element = mem::replace(&mut self.element, MaybeUninit::uninit());
             // SAFETY: We just set count to zero so it won't be dropped again,
             // and it used to be non-zero so it hasn't already been dropped.
-            unsafe { Some(ManuallyDrop::take(&mut self.element)) }
+            unsafe { Some(element.assume_init()) }
```

Comment:
Is this the same as returning `Some(self.element.assume_init_read())`?

Reply:
Hmm. not too sure about this.

Reply:
Writing `uninit()` into  `self.element` will likely be elided in release mode, but explicitly writing `uninit()` into `self.element` would let sanitizers (e.g. Miri) explicitly catch if the field is later read as initialized, even if `T: Copy`.

Compare these two functions: https://rust.godbolt.org/z/cK48E76z3 At `-Copt-level=1`, they are almost the same (notably the `mem::replace(&mut self.element, uninit())` does not actually write anything to `self.element`), and at `-Copt-level=2` or above, the two functions are compiled to the same code.